sql/core/src/test/scala/org/apache/spark/sql/connector/functions/V2FunctionBenchmark.scala

```diff
@@ -0,0 +1,147 @@
+ *      Results will be written to "benchmarks/V2FunctionBenchmark-results.txt".
+ * }}}
+ * '''NOTE''': to update the result of this benchmark, please use Github benchmark action:
+ *   https://spark.apache.org/developer-tools.html#github-workflow-benchmarks
+ */
+object V2FunctionBenchmark extends SqlBasedBenchmark {
+  val catalogName: String = "benchmark_catalog"
+
+  override def runBenchmarkSuite(mainArgs: Array[String]): Unit = {
+    val N = 500L * 1000 * 1000
+    Seq(true, false).foreach { codegenEnabled =>
+      Seq(true, false).foreach { resultNullable =>
+        scalarFunctionBenchmark(N, codegenEnabled = codegenEnabled,
+          resultNullable = resultNullable)
+      }
+    }
+  }
+
+  private def scalarFunctionBenchmark(
+      N: Long,
+      codegenEnabled: Boolean,
+      resultNullable: Boolean): Unit = {
+    withSQLConf(s"spark.sql.catalog.$catalogName" -> classOf[InMemoryCatalog].getName) {
+      createFunction("java_long_add_default",
+        new JavaLongAdd(new JavaLongAddDefault(resultNullable)))
+      createFunction("java_long_add_magic", new JavaLongAdd(new JavaLongAddMagic(resultNullable)))
+      createFunction("java_long_add_static_magic",
+        new JavaLongAdd(new JavaLongAddStaticMagic(resultNullable)))
+      createFunction("scala_long_add_default",
+        LongAddUnbound(new LongAddWithProduceResult(resultNullable)))
+      createFunction("scala_long_add_magic", LongAddUnbound(new LongAddWithMagic(resultNullable)))
+
+      val codeGenFactoryMode = if (codegenEnabled) FALLBACK else NO_CODEGEN
+      withSQLConf(SQLConf.WHOLESTAGE_CODEGEN_ENABLED.key -> codegenEnabled.toString,
+          SQLConf.CODEGEN_FACTORY_MODE.key -> codeGenFactoryMode.toString) {
+        val name = s"scalar function (long + long) -> long, result_nullable = $resultNullable " +
+            s"codegen = $codegenEnabled"
+        val benchmark = new Benchmark(name, N, output = output)
+        benchmark.addCase(s"native_long_add", numIters = 3) { _ =>
+          spark.range(N).select(Column(NativeAdd($"id".expr, $"id".expr, resultNullable))).noop()
+        }
+        Seq("java_long_add_default", "java_long_add_magic", "java_long_add_static_magic",
+            "scala_long_add_default", "scala_long_add_magic").foreach { functionName =>
+          benchmark.addCase(s"$functionName", numIters = 3) { _ =>
+            spark.range(N).selectExpr(s"$catalogName.$functionName(id, id)").noop()
+          }
+        }
+        benchmark.run()
+      }
+    }
+  }
+
+  private def createFunction(name: String, fn: UnboundFunction): Unit = {
+    val catalog = spark.sessionState.catalogManager.catalog(catalogName)
+    val ident = Identifier.of(Array.empty, name)
+    catalog.asInstanceOf[InMemoryCatalog].createFunction(ident, fn)
+  }
+
+  case class NativeAdd(
+      left: Expression,
+      right: Expression,
+      override val nullable: Boolean) extends BinaryArithmetic {
+    override protected val failOnError: Boolean = true
```

Comment:
I think it's better to set it to false. The UDF just does `left + right`, while this native add is doing `math.addExact(left, right)`, which is a bit unfair.

Reply:
Oops you are right. I didn't realize this could impact performance too. Let me do it in a follow-up.